README.md
# MultiLevel Drawer

An easy to implement Multi Level Drawer for Flutter Applications. Just use this in place of regular Scaffold Drawer and you are ready to go.

## Current Features

* Add MultiLevel Drawer to Selected Items
* Change Color of Main and Sub menu According to your Application Needs.

## Demo
![](https://github.com/retroportalstudio/multilevel_drawer/blob/master/multilevel_menu_example.gif)

## Usage
To Use, simply add MultiLevelDrawer to your Scaffold's drawer property:
```
      child: Scaffold(
        drawer: MultiLevelDrawer(
          backgroundColor: Colors.white,
          rippleColor: Colors.white,
          subMenuBackgroundColor: Colors.grey.shade100,
          header: Container(                  // Header for Drawer
            height: size.height * 0.25,
            child: Center(child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Image.asset("assets/dp_default.png",width: 100,height: 100,),
                SizedBox(height: 10,),
                Text("RetroPortal Studio")
              ],
            )),
          ),

          children: [           // Child Elements for Each Drawer Item
            MLMenuItem(
                leading: Icon(Icons.person),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "My Profile",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.settings),
                trailing: Icon(Icons.arrow_right),
                content: Text("Settings"),
                onClick: () {},
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2"))
                ]),
            MLMenuItem(
              leading: Icon(Icons.notifications),
              content: Text("Notifications"),
              onClick: () {},
            ),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "Payments",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
          ],
        ),
        appBar: AppBar(
          backgroundColor: Colors.white,
          iconTheme: IconThemeData(color: Colors.black),
          title: Text(
            "Multi Level Menu",
            style: TextStyle(color: Colors.black),
          ),
        ),
        body: Container(
            decoration: BoxDecoration(
              gradient:
              LinearGradient(begin: Alignment.topCenter, end: Alignment.bottomCenter, colors: [
                Color.fromRGBO(255, 65, 108, 1.0),
                Color.fromRGBO(255, 75, 43, 1.0),
              ]),
            ),
            child: Center()),
      ),
```

Important Classes:
```
MLMenuItem //For Each direct child item of Drawer
MLSubmenu  //For Adding Submenu to MLMenuItem
```

## Roadmap
Plans to add more customizations.

## License
[MIT](https://choosealicense.com/licenses/mit/)

CHANGELOG.md
## [1.0.0] - April 19, 2020

* Add MultiLevel Drawer to Selected Items
* Change Color of Main and Sub menu According to your Application Needs.

## [1.0.1] - April 19, 2020

* Minor Bug Fixes

## [1.0.2] - April 19, 2020

* Minor Bug Fixes

## [1.0.3] - May 24, 2020

* Fixed Navigation Issue with Menu Items

## [1.1.0] - July 03, 2021

* Null Safety
* Drawer Items Scroll
* Adjusted Rendering for SubMenu for Lower Menu Items
* Customizable Item Height


lib/multilevel_drawer.dart
library multilevel_drawer;
import 'package:flutter/material.dart';

class MultiLevelDrawer extends StatefulWidget {
  final Widget header;
  final List<MLMenuItem> children;
  final Color? backgroundColor;
  final Color? subMenuBackgroundColor;
  final Color? divisionColor;
  final LinearGradient? gradient;
  final Color? rippleColor;
  final double? itemHeight;

  const MultiLevelDrawer(
      {required this.header,
      required this.children,
      this.backgroundColor = Colors.white,
      this.gradient,
      this.divisionColor = Colors.grey,
      this.rippleColor = Colors.grey,
      this.subMenuBackgroundColor = Colors.white,
      this.itemHeight = 70});

  @override
  _MultiLevelDrawerState createState() => _MultiLevelDrawerState();
}

class _MultiLevelDrawerState extends State<MultiLevelDrawer> {
  ScrollController scrollController = ScrollController();
  GlobalKey globalKey = GlobalKey();
  List<double> positions = [];
  List<MLMenuItem> drawerItems = [];
  double itemHeight = 0;

  int selectedPosition = -1, lastPosition = 0;
  bool openSubMenu = false;

  @override
  void dispose() {
    super.dispose();
    scrollController.dispose();
  }

  @override
  void initState() {
    itemHeight = widget.itemHeight!;
    positions = [0, 0, 0, 0];
    drawerItems = widget.children;
    super.initState();
    WidgetsBinding.instance!.addPostFrameCallback(getPositions);
    scrollController.addListener(() {
      selectedPosition = -1;
      getPositions(Duration(milliseconds: 0));
    });
  }

  getPositions(duration) {
    RenderBox? renderBox = globalKey.currentContext!.findRenderObject() as RenderBox;
    double dy = renderBox.localToGlobal(Offset.zero).dy;
    double start = dy - 0;
    double end = renderBox.size.height + start;
    double step = itemHeight;
    positions = [];
    for (double x = start; x < end; x = x + step) {
      positions.add(x);
    }
    setState(() {});
  }

  openSubDrawer(position) {
    bool isSamePosition = selectedPosition == position;
    setState(() {
      lastPosition = selectedPosition != -1 ? selectedPosition : 0;
      selectedPosition = isSamePosition ? -1 : position;
    });
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    bool drawUp = false;
    double boxHeight = 0.0;
    if(selectedPosition != -1){
      boxHeight = itemHeight * drawerItems[selectedPosition].subMenuItems!.length;
      drawUp = size.height - positions[selectedPosition] < boxHeight;
    }
    return Container(
      width: size.width,
      child: Stack(
        children: <Widget>[
          if (selectedPosition != -1 && drawerItems[selectedPosition].subMenuItems != null) ...[
            AnimatedPositioned(
                duration: Duration(milliseconds: 200),
                left: selectedPosition == -1 ? 0 : size.width / 2,
                top: selectedPosition != -1 ? (positions[selectedPosition] - (drawUp?(boxHeight - itemHeight):0)) : positions[lastPosition],
                child: Stack(
                  children: <Widget>[
                    Container(
                      width: size.width / 2,
                      height: itemHeight * drawerItems[selectedPosition].subMenuItems!.length,
                    ),
                    Positioned(
                      top: 0,
                      left: 10,
                      child: Container(
                        width: size.width / 2 - 10,
                        child: Column(
                          children: drawerItems[selectedPosition].subMenuItems!.map<Widget>((subMenuItem) {
                            return _MLChoiceItem(
                              height:itemHeight,
                              color: widget.subMenuBackgroundColor,
                              divisionColor: widget.divisionColor,
                              rippleColor: widget.rippleColor,
                              onTap: () {
                                subMenuItem.onClick();
                              },
                              child: subMenuItem.submenuContent,
                            );
                          }).toList(),
                        ),
                      ),
                    ),
                    Positioned(
                        top: drawUp?(boxHeight-itemHeight):0,
                        left: 0,
                        child: CustomPaint(
                          size: Size(10, itemHeight),
                          painter: _ArrowPainter(arrowColor: widget.subMenuBackgroundColor),
                        ))
                  ],
                )),
          ],
          Container(
            decoration: BoxDecoration(color: widget.backgroundColor, gradient: widget.gradient),
            width: size.width / 2,
            height: size.height,
            child: Column(
              children: <Widget>[
                widget.header,
                Expanded(
                  flex: 7,
                  child: SingleChildScrollView(
                    controller: scrollController,
                    child: Container(
                      key: globalKey,
                      height: drawerItems.length * 70,
                      child: ListView.builder(
                        itemCount: drawerItems.length,
                        shrinkWrap: true,
                        primary: false,
                        itemBuilder: (BuildContext context, int index) {
                          MLMenuItem item = drawerItems[index];
                          return _MLChoiceItem(
                            leading: item.leading,
                            trailing: item.trailing,
                            width: size.width / 2,
                            height: itemHeight,
                            divisionColor: widget.divisionColor,
                            color: widget.gradient == null ? widget.backgroundColor : Colors.transparent,
                            rippleColor: widget.rippleColor,
                            child: item.content,
                            onTap: () {
                              if (item.subMenuItems != null) {
                                openSubDrawer(index);
                              } else {
                                item.onClick();
                                openSubDrawer(selectedPosition);
                              }
                            },
                          );
                        },
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class _MLChoiceItem extends StatelessWidget {
  final Function() onTap;
  final Widget? leading;
  final Widget child;
  final Widget? trailing;
  final Color? color;
  final Color? rippleColor;
  final Color? divisionColor;
  final double? width,height;

  const _MLChoiceItem(
      {required this.onTap, required this.child, this.leading, this.trailing, this.color = Colors.white, this.rippleColor = Colors.grey, this.divisionColor = Colors.grey, this.width = 0.0,this.height = 0.0});

  @override
  Widget build(BuildContext context) {
    return Material(
      color: color ?? Colors.white,
      child: InkWell(
        highlightColor: rippleColor,
        onTap: onTap,
        child: Container(
          width: double.infinity,
          height: height,
          decoration: BoxDecoration(border: Border(bottom: BorderSide(color: divisionColor!, width: 1.0))),
          padding: const EdgeInsets.symmetric(vertical: 20, horizontal: 5),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.start,
            children: <Widget>[
              if (leading != null) ...[
                Container(
                  child: leading,
                ),
              ],
              Expanded(child: child),
              if (trailing != null) ...[
                Container(
                  child: trailing,
                ),
              ],
            ],
          ),
        ),
      ),
    );
  }
}

class _ArrowPainter extends CustomPainter {
  final Color? arrowColor;

  _ArrowPainter({this.arrowColor = Colors.white});

  @override
  void paint(Canvas canvas, Size size) {
    Paint paint = Paint()
      ..color = arrowColor!.withOpacity(0.7)
      ..style = PaintingStyle.fill;
    Path path = Path();
    path.moveTo(0, size.height / 2);
    path.lineTo(size.width, 0);
    path.lineTo(size.width, size.height);
    path.close();
    canvas.drawPath(path, paint);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    // TODO: implement shouldRepaint
    return true;
  }
}

class MLMenuItem {
  final Widget content;
  final Widget? leading, trailing;
  final Function onClick;
  final List<MLSubmenu>? subMenuItems;

  const MLMenuItem({required this.content, required this.onClick, this.subMenuItems, this.leading, this.trailing});
}

class MLSubmenu {
  final Widget submenuContent;
  final Function onClick;

  MLSubmenu({required this.submenuContent, required this.onClick});
}


example/lib/main.dart
import 'package:example/SecondScreen.dart';
import 'package:flutter/material.dart';
import 'package:multilevel_drawer/multilevel_drawer.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter ML Menu',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    return SafeArea(
      child: Scaffold(
        drawer: MultiLevelDrawer(
          backgroundColor: Colors.white,
          rippleColor: Colors.white,
          subMenuBackgroundColor: Colors.grey.shade100,
          divisionColor: Colors.grey,
          header: Container(
            height: size.height * 0.25,
            child: Center(
                child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Image.asset(
                  "assets/dp_default.png",
                  width: 100,
                  height: 100,
                ),
                SizedBox(
                  height: 10,
                ),
                Text("RetroPortal Studio")
              ],
            )),
          ),
          children: [
            MLMenuItem(
                leading: Icon(Icons.person),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "My Profile",
                ),
                subMenuItems: [
                  MLSubmenu(
                      onClick: () {
                        Navigator.of(context).pop();
                        Navigator.of(context).push(MaterialPageRoute(builder: (context) => SecondScreen()));
                      },
                      submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 5")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 6")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.settings),
                trailing: Icon(Icons.arrow_right),
                content: Text("Settings"),
                onClick: () {},
                subMenuItems: [MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")), MLSubmenu(onClick: () {}, submenuContent: Text("Option 2"))]),
            MLMenuItem(
              leading: Icon(Icons.notifications),
              content: Text("Notifications"),
              onClick: () {
                Navigator.of(context).push(MaterialPageRoute(builder: (context) => SecondScreen()));
              },
            ),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "Payments",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "Payments",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "Payments",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "New Option",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "New Option 1",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "New Option 2",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "New Option 3",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "New Option 4",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "New Option 5",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 1")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {}),
            MLMenuItem(
                leading: Icon(Icons.payment),
                trailing: Icon(Icons.arrow_right),
                content: Text(
                  "New Option 6",
                ),
                subMenuItems: [
                  MLSubmenu(onClick: () {

                    Navigator.of(context).push(MaterialPageRoute(builder: (context)=>SecondScreen()));
                  }, submenuContent: Text("Option 1")),

                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 2")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 3")),
                  MLSubmenu(onClick: () {}, submenuContent: Text("Option 4")),
                ],
                onClick: () {
                }),
          ],
        ),
        appBar: AppBar(
          backgroundColor: Colors.white,
          iconTheme: IconThemeData(color: Colors.black),
          title: Text(
            "Multi Level Menu",
            style: TextStyle(color: Colors.black),
          ),
        ),
        body: Container(
            decoration: BoxDecoration(
              gradient: LinearGradient(begin: Alignment.topCenter, end: Alignment.bottomCenter, colors: [
                Color.fromRGBO(255, 65, 108, 1.0),
                Color.fromRGBO(255, 75, 43, 1.0),
              ]),
            ),
            child: Center()),
      ),
    );
  }
}
